nav2_msgs/srv/LoadMap.srv

```diff
@@ -1,21 +1,13 @@
-# Type definitions
-uint8 TYPE_FILE=0
-uint8 TYPE_URL=1
-
-# Type of map resource
-uint8 type
-
-# ID of map resource
-string map_id
-
+# URL of map resource
+# Can be an absolute path to a file: file:///path/to/maps/floor1.yaml
+# Or, relative to a ROS package: package://my_ros_package/maps/floor2.yaml
+string map_url
```

Comment:
This should also still take a relative or full path by default if no `[thing]://` exists for backwards compatibility.

That would be a **huge** hindrance for anyone moving to ROS2 to have to change all their maps.

Reply:
My understanding was that this feature doesn't exist in ROS 1 ([yet](https://github.com/ros-planning/navigation/pull/461)), so what do we have to be backwards compatible with?

Reply:
Ah, yes. Sorry I pointed out this line because of the instructions for `[thing]://` not the `.srv` file in particular. I wanted to make sure that in all cases that the `map.yaml` file that contains the ` image: testmap.png` field will accept `image: testmap.png` as well as ` image: file:///path/to/testmap.png ` and ` image: package://my_map_package/maps/testmap.png ` 

Your question made me realize maybe I'm misintepreting this - this is just for the loading from remote servers of the... yaml file? How does the png/jpeg/pgm get picked up? It reads to me like the `read_to_temp` only grabs the `yaml` file, not the `image` file inside of it too. 

I think this PR has gotten to complex for at least me. I think the load map service should be introduced before we start messing around with remote files or changing how we formulate the paths. It seems that if you're going to accept `[thing]://` anywhere, it should be in the `image: [image-name].[image-extension]`, or at least doing that as well when you get the `yaml` file from remote, getting its matched `image` file too

Reply:
Steve, you bring up good points and I think I've addressed them in this PR (which is why it took longer than I hoped).
 - if "package", "http" or "file" are not the first words in the path given, resource_retriever is not used, instead it is treated as an absolute local file path (same as before, so we stay compatible with existing files)
 - if any of those are given, it will fetch the yaml file, parse it, and then treat the image path as a relative path to the yaml file. So if the yaml is at "http://my.server/file.yaml" and the image name given is "file.png" it will be treated as "http://my.server/file.png".

I didn't test the "package" version, but I did test the "http", "file" and default. I've added unit tests for all of them.

Reply:
For bullet 1: only absolute file path? Originally it worked with relative paths and from looking at the code I don’t see in particular why relative would now fail. 

For bullet 2: doesn’t that imply it ignores the yaml key for the image name and just assumes it has the same name in the same directory? That’s not right...

Reply:
bullet 1 - no change, it still works the same as before, you're right that a relative path works. I was thinking only absolute paths do
bullet 2 - it assumes the image file is in the same path as the yaml file, this is unchanged as well. It looks at the file name from the image key in the yaml file, but generates the full path based on the yaml file path. **I didn't change the image key decode logic, it works the same as before.**

The logic that downloads the image file is here:
https://github.com/ros-planning/navigation2/blob/f38818c7ebe095ad326ba199f16e3396841e9e40/nav2_map_server/src/occ_grid_loader.cpp#L313

Reply:
> it assumes the image file is in the same path as the yaml file

There is no assumption of that in the yaml file for the original normal loading. You can specify a full absolute file path from `/home/...` which is not next toe the yaml file. Also relative files don't even have to be in the same filepath, you can have the image file in any number of nested directories.

Reply:
I think you're correct that the logic for loading the image file will have to change.

For retrieving from a locally installed package (`package://`), if an absolute path is given we can assume that it's a path on the local system. If a relative path is given, then we should assume it is relative to the package install directory (eg. we should get the package install prefix with ament_index or the resource_retriever to get the full path to the image file).

For retrieving the map remotely (`http://`), presumably we also need to fetch the image file from the same remote location. This is a bit trickier. If the image path is relative we should swap the last part of the yaml URL with the image path. If the image path is absolute, then we should replace everything after the host name with the path.

For `file://` nothing changes.

Alternatively, we could also change the semantics of the `image` key in the yaml file to also accept a URL, but I think this may be best left for a future iteration to keep things simpler.

Examples:

YAML URL: `package://my_package/path/to/map.yaml`
image key: `relative/path/to/map.png`
Retrieve image here: `package://my_package/path/to/relative/path/to/map.png`

YAML URL: `http://my.remote.host/path/to/map.yaml`
image key: `relative/path/to/map.png`
Retrieve image here: `http://my.remote.host/path/to/relative/path/to/map.png`

YAML URL: `http://my.remote.host/path/to/map.yaml`
image key: `/absolute/path/to/map.png`
Retrieve image here: `http://my.remote.host/absolute/path/to/map.png`

Does this make sense?

Reply:
I would disagree, they have to be changable. Just because the yaml file is being retrieved locally or on a server doesn't mean that the image file will be described the same way. Ei, it shouldn't be on the _resource itself_ to know where the caller is (remote, local, absolute, relative). The yaml doesn't have to be anywhere near the image file, or like in your example 2 its gotten from remote but the image key is by a `package` path and what happens if that package isn't installed on the caller's machine? 

Also we need to be caching these maps, we cannot assume that for every time the server is called that its going to have internet access, if its called it should be downloaded and stored locally so that it can be retrieved locally before trying to call out to a remote server. You don't want to be trying to change maps in the middle of a warehouse in a dead zone and then sit there forever because you have no connection

Reply:
This has alot of edge cases and problems needing to be thought about more. I think all the webstuff should be stripped out of this PR, its trying to change too much at once

Reply:
> Just because the yaml file is being retrieved locally or on a server doesn't mean that the image file will be described the same way

True. I was suggesting we make this assumption for simplicity. As I suggested, we can instead use a URL for the image key, then it doesn't matter where the yaml was retrieved from. There are still edge cases if we **also** allow regular paths (e.g. do we interpret them as being local or relative to the yaml). Perhaps, removing any assumptions about locality between the yaml and image makes things easier.

> in your example 2 its gotten from remote but the image key is by a package path and what happens if that package isn't installed on the caller's machine?

My second example doesn't show the image key is a package path. It is a path relative to map.yaml.

> Also we need to be caching these maps, we cannot assume that for every time the server is called that its going to have internet access, if its called it should be downloaded and stored locally so that it can be retrieved locally before trying to call out to a remote server.

This sounds like an additional feature to be added to the map server and beside the question of what the semantics are for the URL and image key.

> This has alot of edge cases and problems needing to be thought about more. I think all the webstuff should be stripped out of this PR, its trying to change too much at once

+1

It's good that we're having this discussion! But, we should probably move it to https://github.com/ros/common_msgs/pull/152. I agree that for the first pass, this PR should be an incremental change (e.g. exposing the service). I just wanted to make sure we define the service interface appropriately for the future.

Reply:
For the srv type, I have nothing to add, I think its fine and I wouldn't change anything about it. I think the question is in the behavior of use. 

> This sounds like an additional feature to be added to the map server and beside the question of what the semantics are for the URL and image key.

I don't think it is, I think its a precondition to anything with non-local retrieval. We cannot have a situation where we're dependent on the internet to continuously download the same 5 assets we downloaded yesterday. 

I think you and I agree now that this should probably be part of another discussion though, whether its on common msgs or in a ticket here (common_msgs is as good a place as any).

Reply:
For discussion about the behavior, it's probably best to create a ticket in this repo since I think it specific to the nav2_map_server.